### PR TITLE
Controller Input handling refactoring

### LINKED ISSUE #1374: Input mapping GUI incorrectly registering button presses [linux, dualshock4]
Steps to recreate:
Settings > Input > Player 1 Configure > Select Input Device (Controller/0 PS4 Controller) > Select A 'unbound'
Whatever button I set A to gets populated into every field as soon as I press the 'unbound' button

I used various other programs to verify that I did not in fact have a sticky button.
I also verified that the issue persists both with the DS4 kernel driver as well as with the Steam controller driver.

### LINKED ISSUE #1941: PS4 Controller wrong mapping
### What's the issue you encountered?

everytime i try to map any button on my ps4 controller it uses that one button for every time i try to set another button
like i press cross for A and every other button you press to assign will be assigned cross automatically </br>
i did not modify Ryujinx at all because i no not a single scrap of c#

### How can the issue be reproduced?

1.start computer
2.run Linux
3.Connect controller
4.run Ryujninx 
5.attempt to set up input with ps4 controller

### Log file

00:00:00.339 |N| Application PrintSystemInfo: Ryujinx Version: 1.0.0-dirty
00:00:00.672 |N| Application PrintSystemInfo: Operating System: Linux 5.10.7-arch1-1 #1 SMP PREEMPT Wed, 13 Jan 2021 12:02:01 +0000 (X64)
00:00:00.672 |N| Application PrintSystemInfo: CPU: Intel(R) Core(TM) i7-9700F CPU @ 3.00GHz
00:00:00.672 |N| Application PrintSystemInfo: Total RAM: 15941 MB
00:00:00.674 |N| Application PrintSystemInfo: Logs Enabled: Info, Warning, Error, Guest, Stub

### Environment?

- Ryujinx Version: 1.0.0-dirty</br>
- System Specs:
  - OS: *Linux 5.10.7-arch1-1 #1 SMP PREEMPT*
  - CPU: *ntel(R) Core(TM) i7-9700F CPU @ 3.00GHz*
  - GPU: *e.g. NVIDIA RTX 2060*
  - RAM: *16GB*
- Applied Mods : No

### Additional context?

Additional info about your environment:</br>
NA

### PR DESCRIPTION
**Pull request by d3ce1t**
This is my first contribution to this amazing project. Take it easy :)

**Changes:**
- Convert button events to signals in range [-1, 1].
- Detect the pressed button by measuring a change in signal (taking Min, Max and Avg into account).
- In those cases in which a trigger is detected as button and axis, axis wins.
- Joysticks require a full movement in its axis to be detected. i.e one of left -> right,  up -> down, right -> left or down -> up).
- Code has been refactored (new ButtonAssigner interface with JoystickButtonAssigner and KeyboardKeyAssigner implementations)

**Caveats:**
- I have tested only with DS3 in Linux (Fedora 32).
- Hats are not tested at all.
- Likely I have broken all the other controllers in other platforms (testing needed).
- Regarding code organization, I don't know which namespace would be more appropiate for the new classes.

**UPDATES:**
-   1 Feb 2021 - Apply feedback from code review and move classes to Ryujinx.Ui.Input namespace. 
- 18 Jan 2021 - Last commits from master are merged.
- 29 Nov 2020 - PR has been updated with a new version. First comment updated too.
- 24 Nov 2020 - Initial version of this PR.

This should fixes #1374 and fixes #1941 in Linux.

### CODE REVIEW

**Comment by d3ce1t** [comment]
BTW looks like the previous code didn't work in Linux because Joystick.GetState was retrieving a wrong state.

After the first button was pressed and bounded to a key, subsequent Joystick.GetState calls were returning a state with that button in "down" state (despite being "up" actually). I don't know OpenTK.Input, but it seems like input is buffered and needs to be consumed completely.

**Comment by qurious-pixel** [comment]
Latest build 1.0.5819 works well on linux with DS4.
i was able to input the controls and save the profile.
Then refreshed and recalled the profile correctly.

**Comment by d3ce1t** [comment]
> Latest build 1.0.5819 works well on linux with DS4.
> i was able to input the controls and save the profile.
> Then refreshed and recalled the profile correctly.

I guess you mean build 1.0.5816, right?

**Comment by qurious-pixel** [comment]
This one.
[build 1.0.5819](https://ci.appveyor.com/project/gdkchan/ryujinx/builds/36493460)

**Comment by d3ce1t** [comment]
Thank you! I didn't know the build of this PR was available there. So I'm glad it worked for you too.

What I'm curious about is whether it works in Windows (I can't test that)

**Comment by EmulationFanatic** [comment]
@d3ce1t Can you rebase this?

**Comment by d3ce1t** [comment]
@EmulationFanatic Done :)

**Comment by EmulationFanatic** [comment]
Tested input mapping with three controllers in Windows across multiple boots of the application to see if there had been any changes compared to master.
-Xbox 360 for Windows controller: no change in operation. Worked as intended.
-Pro Controller with BetterJoy: no change in operation. Worked as intended.
-Logitech Dual Action (Dinput controller): this PR resolved the issue with Button12 being spammed while trying to map inputs. I was able to map all buttons/axes without any trouble this time.

Big thumbs up from me!

**Comment by Doctorwho1909** [comment]
this fixes #1941 also

File: Ryujinx/Ui/Windows/ControllerWindow.cs [hunk 1]
```diff
@@ -688,74 +621,66 @@ private void RefreshInputDevicesButton_Pressed(object sender, EventArgs args)
             _refreshInputDevicesButton.SetStateFlags(StateFlags.Normal, true);
         }
 
+        private ButtonAssigner CreateButtonAssigner()
+        {
+            int index = int.Parse(_inputDevice.ActiveId.Split("/")[1]);
+
+            ButtonAssigner assigner;
+
+            if (_inputDevice.ActiveId.StartsWith("keyboard")) {
```
Comment (AcK77) [inline Ryujinx/Ui/Windows/ControllerWindow.cs:630]: Add the bracket on a new line

File: Ryujinx/Ui/Windows/ControllerWindow.cs [hunk 2]
```diff
@@ -688,74 +621,66 @@ private void RefreshInputDevicesButton_Pressed(object sender, EventArgs args)
...
+                assigner = new KeyboardKeyAssigner(index);
+            }
+            else if (_inputDevice.ActiveId.StartsWith("controller")) {
```
Comment (AcK77) [inline Ryujinx/Ui/Windows/ControllerWindow.cs:633]: Add the bracket on a new line

File: Ryujinx/Ui/Windows/ControllerWindow.cs [hunk 3]
```diff
@@ -688,74 +621,66 @@ private void RefreshInputDevicesButton_Pressed(object sender, EventArgs args)
...
+                // TODO: triggerThresold is passed but not used by JoystickButtonAssigner. Should it be used for key binding?.
+                // Note that, like left and right sticks, ZL and ZR triggers are treated as axis.
+                // The problem is then how to decide which axis should use triggerThresold.
+                assigner = new JoystickButtonAssigner(index, _controllerTriggerThreshold.Value);
+            }
+            else {
```
Comment (AcK77) [inline Ryujinx/Ui/Windows/ControllerWindow.cs:639]: Add the bracket on a new line

File: Ryujinx/Ui/Windows/ControllerWindow.cs [hunk 4]
```diff
@@ -688,74 +621,66 @@ private void RefreshInputDevicesButton_Pressed(object sender, EventArgs args)
...
-                                button.SetStateFlags(StateFlags.Normal, true);
-                            });
-
-                            _isWaitingForInput = false;
-
-                            return;
-                        }
+                    Thread.Sleep(10);
+                    assigner.ReadInput();
+                    if (assigner.HasAnyButtonPressed() || assigner.ShouldCancel()) {
```
Comment (AcK77) [inline Ryujinx/Ui/Windows/ControllerWindow.cs:665]: Add an empty line before and add the bracket on a new line

File: Ryujinx/Ui/Windows/ControllerWindow.cs [hunk 5]
```diff
@@ -688,74 +621,66 @@ private void RefreshInputDevicesButton_Pressed(object sender, EventArgs args)
...
+                var pressedButton = assigner.GetPressedButton();
 
-                            _isWaitingForInput = false;
+                ToggleButton button = (ToggleButton) sender;
 
-                            return;
-                        }
+                Application.Invoke(delegate
+                {
+                    if (pressedButton != "") {
```
Comment (AcK77) [inline Ryujinx/Ui/Windows/ControllerWindow.cs:676]: Add the bracket on a new line

**Review by AcK77 — CHANGES_REQUESTED** [review]
It could be nice to move the files in the `UI` folder to another folder, maybe in the `Helper` folder or another one which contains all classes/files related to input.

**Review by Xpl0itR — APPROVED** [review]

**Comment by Xpl0itR** [comment]
nit: Use explicit types instead of var, other than that lgtm

**Comment by d3ce1t** [comment]
Thanks for your feedback @Xpl0itR. Changed :)

File: Ryujinx/Ui/Input/JoystickButtonAssigner.cs [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+            return pressedButtons.Count > 0 ? pressedButtons[0].ToString() : "";
+        }
+
+        private void CollectButtonStats()
+        {
+            JoystickCapabilities capabilities = Joystick.GetCapabilities(_index);
+
+            ControllerInputId pressedButton;
+
+            //Buttons
```
Comment (Thog) [inline Ryujinx/Ui/Input/JoystickButtonAssigner.cs:67]:
```suggestion
            // Buttons
```

File: Ryujinx/Ui/Input/JoystickButtonAssigner.cs [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+                }
+
+                if (_currState.IsButtonUp(i) && _prevState.IsButtonDown(i))
+                {
+                    Enum.TryParse($"Button{i}", out pressedButton);
+                    _detector.AddInput(pressedButton, -1);
+                }
+            }
+
+            //Axis
```
Comment (Thog) [inline Ryujinx/Ui/Input/JoystickButtonAssigner.cs:83]:
```suggestion
            // Axis
```

File: Ryujinx/Ui/Input/JoystickButtonAssigner.cs [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+            for (int i = 0; i != capabilities.AxisCount; i++)
+            {
+                float axisValue = _currState.GetAxis(i);
+
+                Enum.TryParse($"Axis{i}", out pressedButton);
+                _detector.AddInput(pressedButton, axisValue);
+            }
+
+            //Hats
```
Comment (Thog) [inline Ryujinx/Ui/Input/JoystickButtonAssigner.cs:92]:
```suggestion
            // Hats
```

Comment (AcK77) [inline Ryujinx/Ui/Input/JoystickButtonAssigner.cs:92]:
```suggestion
            // Hats
```

File: Ryujinx/Ui/Input/JoystickButtonAssigner.cs (hunk 7 above)
Comment (AcK77) [inline Ryujinx/Ui/Input/JoystickButtonAssigner.cs:83]:
```suggestion
            // Axis
```

File: Ryujinx/Ui/Input/JoystickButtonAssigner.cs (hunk 6 above)
Comment (AcK77) [inline Ryujinx/Ui/Input/JoystickButtonAssigner.cs:67]:
```suggestion
            // Buttons
```

**Review by Thog — APPROVED** [review]
lgtm, just a few nits

File: Ryujinx/Ui/Input/JoystickButtonAssigner.cs [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+        public bool ShouldCancel()
+        {
+            return Mouse.GetState().IsAnyButtonDown || Keyboard.GetState().IsAnyKeyDown;
+        }
+
+        public string GetPressedButton()
+        {
+            List<ControllerInputId> pressedButtons = _detector.GetPressedButtons();
+
+            // Reverse list so axis button take precedence when more than one button is recognized
```
Comment (AcK77) [inline Ryujinx/Ui/Input/JoystickButtonAssigner.cs:55]:
```suggestion
            // Reverse list so axis button take precedence when more than one button is recognized.
```

File: Ryujinx/Ui/Input/JoystickButtonAssigner.cs [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+        {
+            _currState = Joystick.GetState(_index);
+            _prevState = _currState;
+        }
+
+        public void ReadInput()
+        {
+            _prevState = _currState;
+            _currState = Joystick.GetState(_index);
+            CollectButtonStats();
```
Comment (AcK77) [inline Ryujinx/Ui/Input/JoystickButtonAssigner.cs:38]: Add a new empty line before

**Review by AcK77 — CHANGES_REQUESTED** [review]

**Comment by AcK77** [comment]
Hi @d3ce1t any news about the reviews?
Thanks!

**Comment by d3ce1t** [comment]
Ey Sorry @AcK77! I cound't apply your changes yet. Very busy these days. Likely this weekend I will have time to do it :)

**Comment by d3ce1t** [comment]
Thanks for your feedback. All changes included :)

**Review by AcK77 — APPROVED** [review]
Lgtm! Thanks!

**Pull request merged**